sympy/tensor/array/expressions/array_expressions.py

```diff
@@ -1571,7 +1571,7 @@ def __init__(self, base_array: typing.Union[ArrayContraction, ArrayDiagonal, Arr
                 args_with_ind[arg_pos].indices[rel_pos] = i
         self.args_with_ind: List[_ArgE] = args_with_ind
         self.number_of_contraction_indices: int = len(contraction_indices)
-        self._track_permutation: Optional[List[List[int]]] = None
+        self._track_permutation: List[List[int]] = None # type: ignore
```

Comment:
Would `List[List[int]] | NoneType` allow the use of `None` here without ignore?

Reply:
Maybe. I'll check...

Reply:
That gives
```console
$ mypy sympy
sympy/tensor/array/expressions/array_expressions.py:1574: error: X | Y syntax for unions requires Python 3.10
sympy/tensor/array/expressions/array_expressions.py:1574: error: Name "NoneType" is not defined
```

Reply:
Reading the docs, it looks like it is exactly what `Optional` does in the first place: https://docs.python.org/3/library/typing.html#typing.Optional 

Why was that changed? It feels like a step backwards.

Reply:
I'm not sure exactly how to type this given that before use it will always become non-None. If you have Optional here then mypy will complain every time this is used:
```
$ mypy sympy
sympy/tensor/array/expressions/conv_array_to_matrix.py:959: error: Value of type "Optional[List[List[int]]]" is not indexable
sympy/tensor/array/expressions/conv_array_to_matrix.py:973: error: Item "None" of "Optional[List[List[int]]]" has no attribute "__iter__" (not iterable)
Found 2 errors in 1 file (checked 1455 source files)
```

Reply:
I see, it seems that mypy simply doesn't allow you to do this unless you handle "None" in the code explicitly everywhere even if you know it will never be `None`.

Reply:
I'll just add back the type: ignore. This class is too complicated for me to figure out how to clean it up. (I don't think it needs to be complicated but the code is really messy)

Reply:
```diff
diff --git a/sympy/tensor/array/expressions/array_expressions.py b/sympy/tensor/array/expressions/array_expressions.py
index 75573cbb32..c42c53babb 100644
--- a/sympy/tensor/array/expressions/array_expressions.py
+++ b/sympy/tensor/array/expressions/array_expressions.py
@@ -1571,7 +1571,7 @@ def __init__(self, base_array: typing.Union[ArrayContraction, ArrayDiagonal, Arr
                 args_with_ind[arg_pos].indices[rel_pos] = i
         self.args_with_ind: List[_ArgE] = args_with_ind
         self.number_of_contraction_indices: int = len(contraction_indices)
-        self._track_permutation: Optional[List[List[int]]] = None
+        self._track_permutation: List[List[int]] = [[]]

         mapping = _get_mapping_from_subranks(base_array.subranks)

@@ -1673,7 +1673,7 @@ def to_array_contraction(self):
         contraction_indices = self.get_contraction_indices()
         expr = _array_contraction(_array_tensor_product(*args), *contraction_indices)
         expr2 = _array_diagonal(expr, *diag_indices_filtered)
-        if self._track_permutation is not None:
+        if any(self._track_permutation):
             permutation2 = _af_invert([j for i in self._track_permutation for j in i])
             expr2 = _permute_dims(expr2, permutation2)

@@ -1755,8 +1755,8 @@ def track_permutation_start(self):
     def track_permutation_merge(self, destination: _ArgE, from_element: _ArgE):
         index_destination = self.args_with_ind.index(destination)
         index_element = self.args_with_ind.index(from_element)
-        self._track_permutation[index_destination].extend(self._track_permutation[index_element]) # type: ignore
-        self._track_permutation.pop(index_element) # type: ignore
+        self._track_permutation[index_destination].extend(self._track_permutation[index_element])
+        self._track_permutation.pop(index_element)

     def get_absolute_free_range(self, arg: _ArgE) -> typing.Tuple[int, int]:
         """
```
It seems to pass `mypy`, `bin/test sympy/tensor`, `bin/doctest/ sympy/tensor`, but haven't done the full tests

Reply:
There is this which might be affected:
https://github.com/sympy/sympy/blob/c43550c31ef6355c49c4461e90a23e9629291b93/sympy/tensor/array/expressions/array_expressions.py#L1676

Reply:
The reason why it has that is because it uses `self._track_permutation` in a list comprehension, so the `any(self._track_permutation)` check should identify the uninitialized list.

Alternative is to use what mypy seems to imply:
```diff
diff --git a/sympy/tensor/array/expressions/array_expressions.py b/sympy/tensor/array/expressions/array_expressions.py
index 75573cbb32..079a1febe6 100644
--- a/sympy/tensor/array/expressions/array_expressions.py
+++ b/sympy/tensor/array/expressions/array_expressions.py
@@ -1755,8 +1755,9 @@ def track_permutation_start(self):
     def track_permutation_merge(self, destination: _ArgE, from_element: _ArgE):
         index_destination = self.args_with_ind.index(destination)
         index_element = self.args_with_ind.index(from_element)
-        self._track_permutation[index_destination].extend(self._track_permutation[index_element]) # type: ignore
-        self._track_permutation.pop(index_element) # type: ignore
+        if self._track_permutation is not None:
+            self._track_permutation[index_destination].extend(self._track_permutation[index_element])
+            self._track_permutation.pop(index_element)

     def get_absolute_free_range(self, arg: _ArgE) -> typing.Tuple[int, int]:
         """
```

Reply:
This code is too messy for me to try and make changes to it. I don't really understand what it does and I don't know how well tested it is so I'd rather just leave the `type: ignore` there and let someone who wants to improve this code fix it. At least I don't want to smuggle a bug in with a PR that has a large number of trivial changes.

Reply:
Ok, fair enough.